elastic/elasticsearch
# Adding cluster alert for ES oversharding

This PR adds a Stack Monitoring cluster alert for a possibly oversharded Elasticsearch cluster. A cluster is considered to be possibly oversharded if it has more than 20 shards / GB of max heap space.

Related: https://github.com/elastic/kibana/pull/37496.

### Testing this PR

1. Make sure you are running Elasticsearch with a Trial license (as cluster alerts require Alerting).
2. Visit the Stack Monitoring UI in Kibana and make sure Stack Monitoring is enabled for the cluster.
3. Initially, make sure you are running fewer than 20 shards / GB of max. heap space. This should be the case with a default install of a single-node ES cluster.
4. Visit the Stack Monitoring UI > Cluster Overview and make sure there are no cluster alerts about having too many shards.
5. Create a test index with too many shards. For example, for a one-node ES cluster with 1 GB of max. heap space, create this index:

   ```
   PUT oversharder
   {
     "settings": {
       "number_of_shards": 20,
       "number_of_replicas": 0
     }
   }
   ```
6. After a minute or so, visit the Stack Monitoring UI > Cluster Overview and check that a cluster alert like this one shows up:

   <img width="1263" alt="Screen Shot 2019-05-29 at 5 26 27 PM" src="https://user-images.githubusercontent.com/51061/58600025-93715100-8237-11e9-897f-1e8ed008b29f.png">
